dbms/src/Storages/IStorage.h

```diff
@@ -345,28 +339,20 @@ class IStorage : public std::enable_shared_from_this<IStorage>, private boost::n
     using std::enable_shared_from_this<IStorage>::shared_from_this;
 
 private:
-    friend class TableStructureReadLock;
+    /// You always need to take the next three locks in this order.
 
-    /// You always need to take the next two locks in this order.
+    /// If you hold this lock exclusively, you can be sure that no other structure modifying queries
+    /// (e.g. ALTER, DROP) are concurrently executing. But queries that only read table structure
+    /// (e.g. SELECT, INSERT) can continue to execute.
+    mutable RWLock alter_intention_lock = RWLockImpl::create();
 
-    /** It is taken for read for the entire INSERT query and the entire merge of the parts (for MergeTree).
-      * It is taken for write for the entire time ALTER MODIFY.
-      *
-      * Formally:
-      * Taking a write lock ensures that:
-      *  1) the data in the table will not change while the lock is alive,
-      *  2) all changes to the data after releasing the lock will be based on the structure of the table at the time after the lock was released.
-      * You need to take for read for the entire time of the operation that changes the data.
-      */
-    mutable RWLock data_lock = RWLockImpl::create();
+    /// It is taken for share for the entire INSERT query and the entire merge of the parts (for MergeTree).
+    /// ALTER queries acquire an exclusive lock to ensure
```

Comment:
To ensure what?

Reply:
... that no new data appear?

Reply:
oops :) I will add the missing part of the comment.